File: contest_practice/silver/snowboots/snowboots.java
import java.io.*;
import java.util.*;
public class snowboots {
    public static void main(String args[]) throws IOException {
        BufferedReader fin = new BufferedReader(new FileReader("snowboots.in"));
        StringTokenizer st = new StringTokenizer(fin.readLine());
        int n = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(fin.readLine());
        int[] tiles = new int[n];
        for (int i = 0; i < n; i++) {
            tiles[i] = Integer.parseInt(st.nextToken());
        }
        int[][] boots = new int[b][2];
        for (int i = 0; i < b; i++) {
            st = new StringTokenizer(fin.readLine());
            boots[i][0] = Integer.parseInt(st.nextToken());
            boots[i][1] = Integer.parseInt(st.nextToken());
        }

        int result = 0;
        int i = 0;
        for (int boot = 0; boot < b; boot++) {
            if (i == n-1) {
                result = boot;
                break;
            }
            for (int j = i + boots[boot][1]; j > i; j--) {
                if (j >= n) {
                    continue;
                }
                if (tiles[j] > boots[boot][0]) {
                    continue;
                }
                i = j;
                break;
            }
        }

        System.out.println(result);
    }
}

File: contest_practice/silver/snowboots/snowboots.py
with open("snowboots.in", "r") as f:
    n, b = [int(x) for x in f.readline().strip().split()]
    tiles = [int(x) for x in f.readline().strip().split()]
    boots = []
    for i in range(b):
        boots.append([int(x) for x in f.readline().strip().split()])

used = [float("inf") for i in range(n)]
used[0] = 0

# print(boots)
# print(tiles)
for i in range(n):
    # print(i, used)
    if used[i] == float("inf"):
        continue
    for j in range(used[i], b):
        if (tiles[i] > boots[j][0]):
            continue
        for x in range(boots[j][1]):
            new_i = i+boots[j][1]-x
            if new_i >= n:
                continue
            if j < used[new_i] and boots[j][0] >= tiles[new_i]:
                used[new_i] = j

# print(used)
with open("snowboots.out", "w") as f:
    f.write(str(used[-1]))
